wgpu-core/src/device/any_device.rs

```diff
@@ -1,72 +1,98 @@
+use wgt::Backend;
+
 use super::Device;
 /// The `AnyDevice` type: a pointer to a `Device<A>` for any backend `A`.
 use crate::hal_api::HalApi;
 
-use std::any::Any;
 use std::fmt;
+use std::mem::ManuallyDrop;
+use std::ptr::NonNull;
 use std::sync::Arc;
 
+struct AnyDeviceVtable {
+    // We oppurtunistically store the backend here, since we now it will be used
+    // with backend selection and it can be stored in static memory.
+    backend: Backend,
+    // Drop glue which knows how to drop the stored data.
+    drop: unsafe fn(*mut ()),
+}
+
 /// A pointer to a `Device<A>`, for any backend `A`.
 ///
-/// Any `AnyDevice` is just like an `Arc<Device<A>>`, except that the
-/// `A` type parameter is erased. To access the `Device`, you must
-/// downcast to a particular backend with the \[`downcast_ref`\] or
-/// \[`downcast_clone`\] methods.
-pub struct AnyDevice(Arc<dyn Any + 'static>);
+/// Any `AnyDevice` is just like an `Arc<Device<A>>`, except that the `A` type
+/// parameter is erased. To access the `Device`, you must downcast to a
+/// particular backend with the \[`downcast_ref`\] or \[`downcast_clone`\]
+/// methods.
+pub struct AnyDevice {
+    data: NonNull<()>,
+    vtable: &'static AnyDeviceVtable,
+}
 
 impl AnyDevice {
     /// Return an `AnyDevice` that holds an owning `Arc` pointer to `device`.
     pub fn new<A: HalApi>(device: Arc<Device<A>>) -> AnyDevice {
-        AnyDevice(device)
+        unsafe fn drop_glue<A: HalApi>(ptr: *mut ()) {
+            // Drop the arc this instance is holding.
+            unsafe {
+                _ = Arc::from_raw(ptr.cast::<A::Surface>());
+            }
```

Comment:
nitpick: I think it would be clearer to use [`Arc::decrement_strong_count`](https://doc.rust-lang.org/stable/std/sync/struct.Arc.html#method.decrement_strong_count) here:

```suggestion
            unsafe {
                Arc::decrement_strong_count(ptr.cast::<A::Surface>());
            }
```